### DISCUSSION TITLE
Live streaming

### DISCUSSION THREAD

**Discussion by 19942214** [raphaelvallat/yasa#56] [Q&A] [answered]
Hi,

I just discovered your page and really impressed with it! 
I don't know if I missed this in the documentation, but will YASA work for live streaming of EEG data? I am trying to do sleep staging with in-ear EEG, so I will only have 4 channels (but I don't think that will be a problem if I have fewer channels than the 19 as in your examples, or would it?) and want to be able to detect the different stages of sleep while the person is sleeping for an action needs to occur once the person is in NREM2. That is why I need the spindle detection function to identify the difference between NREM1 and NREM2.

Would it be possible to use YASA for live streaming of EEG?

Let me know if I was unclear with my description. Looking forward to hearing from you.

**Comment by raphaelvallat** [comment, marked as answer]
Hi @19942214 — great question. YASA was only designed for offline uses, and therefore will not work for real-time scoring. That said, if you can get access to the real-time + historical data (e.g. all the data that has been recorded so far) in let's say a numpy array or MNE object, it is technically feasible to just repeatedly call YASA every new epoch and extract the sleep stages up until this point. However, this won't be perfectly real-time, because it can only be done once the current 30-sec epoch is over.

Hope this helps,

Thanks,
Raphael

**Reply by 19942214** [reply]
Thanks for your reply! 

I understand what you are saying. I was thinking of something like that in case you said that it would not be possible. My project is a proof of concept so not working with 100% real-time should not be an issue. Like you said, waiting for the first 30s epoch to pass, evaluating it, and waiting for the next 30s epoch to be evaluated next.

I am working with the openBCI's Cyton board with python to do the processing and started with Brainflow to set up the initialization of the board before I saw your work. So I was thinking to integrate both Brainflow and your work together (obviously citing your work in the code) and get the Cyton board to do the sleep staging. Do you think something like that would be possible? I saw Brainflow does have an example code for Python MNE integration which I am looking into to get the data in the correct format for your code. Here is a link if you want to look: https://brainflow.readthedocs.io/en/stable/Examples.html#python-mne-integration

Thank you for all your help and advice!
Elke

**Reply by 19942214** [reply]
Hello again,

I am trying to use the Automatic Sleep Staging Jupyter Notebook example on your Github page (https://github.com/raphaelvallat/yasa/blob/master/notebooks/14_automatic_sleep_staging.ipynb) and also trying to implement as we discussed the waiting for 30s epoch to be evaluated for sleep staging. I get the error saying that the SleepStaging() function needs at least 5 minutes of data to work. Is there a way around this?

Kind regards,
Elke

**Reply by raphaelvallat** [reply]
Hi @19942214,

Unfortunately this is hardcoded into the algorithm so unless you modify the actual code there is no way around it. Unless you do wait to have at least 5 min of EEG data (can also include pre-sleep wakefulness) before applying any scoring? Remember that YASA will not be able to score only a single isolated 30-sec epoch of data, you will need to have some data from the past epochs as well.

I have never worked with Brainflow or OpenBCI, but looking at the code snippet that you share I do think it should be possible to integrate with YASA.

Best of luck,

Thanks,
Raphael

**Reply by 19942214** [reply]
Hi, 

Okay, that is fine. Just to confirm, after the first 5 min of plain EEG recording YASA will then be able to score the 30s epoch of data? Or should I do: firstly record the 5 min minimum of EEG data and score that and then just append the 30s epochs to that array for YASA to evaluate? Will something like that work? Then YASA will have the 5 min minimum of data to record and each time 30-sec more to score? 

Then another question, in your automatic sleep scoring example you used:

# Let's now load the human-scored hypnogram, where each value represents a 30-sec epoch.
hypno = np.loadtxt('sub-02_hypno_30s.txt', dtype=str)
hypno

Is the hypnogram only used to test the accuracy of my collected data as shown later on in the example? Also just want to confirm that I am understanding it correctly.

Thanks for your patience and for answering my questions,
Elke

**Reply by raphaelvallat** [reply]
Hi @19942214,

You wait to have at least 5 min of data and then apply YASA to get the sleep stages. After each additional 30-sec of recorded data, you re-apply the sleep staging algorithm on the entire dataset (i.e. 5 min, then 5min30, then 6 min). YASA will become more and more accurate as you record more data. Please note that you might not get exactly the same hypnogram with 5min or 5min30 of data, since YASA uses past data to determine the current epoch, therefore, the more data you have, the better.

Yes, in this example the hypnogram is loaded to compare YASA against the ground-truth (manual) scoring.

Thanks,
Raphael

**Comment by 19942214** [comment]
Thanks for the advice!

What does YASA's processing time look like for sleep data of an hour, or maybe even 5 hours? Does the processing time decrease a lot when there is for example 5 hours of sleep data to be processed? Just wondering if I would need to clear some of the earlier data in the raw array if the processing time is starting to take too long?

Thanks,
Elke

**Reply by 19942214** [reply]
Also with the SleepStaging() function, can the function only do it for 1 channel or multiple? I want to get the sleep staging for all four channels, but I get an error that it cannot define the variable if I give the argument eeg_name = "Fp1-C4" as the description of the function states (- "eeg_name" is the name of the EEG channel, preferentially a central derivation (e.g. C4-M1). This is always required to run the sleep staging algorithm.) 

Am I typing it wrong or is there another way for it to define it? 

Sorry for all the questions.

Thanks,
Elke

**Reply by raphaelvallat** [reply]
Hi @19942214,

Processing time should be between 2-20 seconds, regardless of the amount of data you have (i.e. roughly similar for 1 hour or 8 hours of data). It will be faster if your data is downsampled at 100 Hz.

YASA only supports one EEG channel for now (preferentially a central derivation, e.g. C4-M1 or C3-M2). See this issue for more details: https://github.com/raphaelvallat/yasa/issues/40

Thanks,
Raphael
